### PR TITLE
[type: feature] add nacos, etcd, eureka discovery service

### PR DESCRIPTION
**Pull request by eurecalulu**
In this PR, I expanded the service discovery functions of nacos, etcd, and eureka registration centers, and passed comprehensive tests.
Make sure that:

- [ ] You have read the [contribution guidelines](https://shenyu.apache.org/community/contributor-guide).
- [ ] You submit test cases (unit or integration tests) that back your changes.
- [ ] Your local test passed `./mvnw clean install -Dmaven.javadoc.skip=true`.

### CODE REVIEW

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/discovery/DefaultDiscoveryProcessor.java [hunk 1]
```diff
@@ -176,6 +176,7 @@ public void fetchAll(final String discoveryHandlerId) {
         String discoveryId = discoveryHandlerDO.getDiscoveryId();
         if (discoveryServiceCache.containsKey(discoveryId)) {
             ShenyuDiscoveryService shenyuDiscoveryService = discoveryServiceCache.get(discoveryId);
+            // TODO 如果列表为空？
```
Comment (moremind) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/discovery/DefaultDiscoveryProcessor.java:179]: remove this

File: shenyu-discovery/shenyu-discovery-api/src/main/java/org/apache/shenyu/discovery/api/ShenyuDiscoveryService.java [hunk 2]
```diff
@@ -62,18 +62,18 @@ public interface ShenyuDiscoveryService {
     /**
      * getData by key.
      *
-     * @param key key
+     * @param key keyd
```
Comment (moremind) [inline shenyu-discovery/shenyu-discovery-api/src/main/java/org/apache/shenyu/discovery/api/ShenyuDiscoveryService.java:65]: key

File: shenyu-discovery/shenyu-discovery-etcd/pom.xml [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+            <artifactId>shenyu-discovery-api</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.shenyu</groupId>
+            <artifactId>shenyu-common</artifactId>
+            <version>${project.version}</version>
+            <scope>compile</scope>
+        </dependency>
+<!--        <dependency>-->
```
Comment (moremind) [inline shenyu-discovery/shenyu-discovery-etcd/pom.xml:42]: remove this

File: shenyu-discovery/shenyu-discovery-etcd/src/main/java/org/apache/shenyu/discovery/etcd/EtcdDiscoveryService.java [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+    private long ttl;
+
+    private long timeout;
+
+    @Override
+    public void init(final DiscoveryConfig config) {
+        Properties props = config.getProps();
+        this.timeout = Long.parseLong(props.getProperty("etcdTimeout", "3000"));
+        this.ttl = Long.parseLong(props.getProperty("etcdTTL", "5"));
+        if (this.etcdClient == null) {
```
Comment (moremind) [inline shenyu-discovery/shenyu-discovery-etcd/src/main/java/org/apache/shenyu/discovery/etcd/EtcdDiscoveryService.java:75]: use Objects.isNull

File: shenyu-discovery/shenyu-discovery-eureka/pom.xml [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+            <scope>compile</scope>
+        </dependency>
+        <dependency>
+            <groupId>com.netflix.eureka</groupId>
+            <artifactId>eureka-client</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>com.alibaba.fastjson2</groupId>
+            <artifactId>fastjson2</artifactId>
+            <version>2.0.14</version>
```
Comment (moremind) [inline shenyu-discovery/shenyu-discovery-eureka/pom.xml:50]: not use fastjson

File: shenyu-discovery/shenyu-discovery-eureka/src/main/java/org/apache/shenyu/discovery/eureka/EurekaDiscoveryService.java [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+            List<InstanceInfo> initialInstances = eurekaClient.getInstancesByVipAddressAndAppName(null, key, true);
+            instanceListMap.put(key, initialInstances);
+            ScheduledFuture<?> scheduledFuture = executorService.scheduleAtFixedRate(() -> {
+                try {
+                    List<InstanceInfo> previousInstances = instanceListMap.get(key);
+                    List<InstanceInfo> currentInstances = eurekaClient.getInstancesByVipAddressAndAppName(null, key, true);
+                    compareInstances(previousInstances, currentInstances, listener);
+                    instanceListMap.put(key, currentInstances);
+                } catch (Exception e) {
+                    throw new ShenyuException(e);
```
Comment (moremind) [inline shenyu-discovery/shenyu-discovery-eureka/src/main/java/org/apache/shenyu/discovery/eureka/EurekaDiscoveryService.java:100]: add Log.error("watch discovery data error=", e)

File: shenyu-discovery/shenyu-discovery-eureka/src/main/java/org/apache/shenyu/discovery/eureka/EurekaDiscoveryService.java [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+            }, 0, 1, TimeUnit.SECONDS);
+            listenerThreadsMap.put(key, scheduledFuture);
+        }
+    }
+
+    @Override
+    public void unwatch(final String key) {
+        try {
+            ScheduledFuture<?> scheduledFuture = listenerThreadsMap.get(key);
+            if (scheduledFuture != null) {
```
Comment (moremind) [inline shenyu-discovery/shenyu-discovery-eureka/src/main/java/org/apache/shenyu/discovery/eureka/EurekaDiscoveryService.java:111]: Objects.isNull

File: shenyu-discovery/shenyu-discovery-nacos/pom.xml [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+        <dependency>
+            <groupId>org.apache.shenyu</groupId>
+            <artifactId>shenyu-common</artifactId>
+            <version>${project.version}</version>
+            <scope>compile</scope>
+        </dependency>
+        <dependency>
+            <groupId>com.alibaba.fastjson2</groupId>
+            <artifactId>fastjson2</artifactId>
+            <version>2.0.14</version>
```
Comment (moremind) [inline shenyu-discovery/shenyu-discovery-nacos/pom.xml:45]: remove this

**Review by moremind — CHANGES_REQUESTED** [review]

File: shenyu-discovery/shenyu-discovery-eureka/src/main/java/org/apache/shenyu/discovery/eureka/EurekaDiscoveryService.java [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+public class EurekaDiscoveryService implements ShenyuDiscoveryService {
+    private static final Logger LOGGER = LoggerFactory.getLogger(EurekaDiscoveryService.class);
+
+    private ApplicationInfoManager applicationInfoManager;
+
+    private EurekaClient eurekaClient;
+
+    private EurekaHttpClient eurekaHttpClient;
+
+    private final ScheduledExecutorService executorService = Executors.newScheduledThreadPool(10);
```
Comment (moremind) [inline shenyu-discovery/shenyu-discovery-eureka/src/main/java/org/apache/shenyu/discovery/eureka/EurekaDiscoveryService.java:67]: not use Executors create thread pool

**Review by moremind — CHANGES_REQUESTED** [review]

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/discovery/DefaultDiscoveryProcessor.java (hunk 1 above)
Comment (eurecalulu) [reply shenyu-admin/src/main/java/org/apache/shenyu/admin/discovery/DefaultDiscoveryProcessor.java:179]: Thank you for your review, I will modify it immediately.

File: pom.xml [hunk 10]
```diff
@@ -197,12 +197,6 @@
                 <scope>import</scope>
             </dependency>
 
-            <dependency>
-                <groupId>com.alibaba.nacos</groupId>
```
Comment (moremind) [inline pom.xml:201]: why remove this? this is dependency management

**Review by moremind — CHANGES_REQUESTED** [review]

**Review by moremind — APPROVED** [review]

**Pull request merged**